llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -6,30 +6,120 @@
 
 #include "llvm/ADT/Triple.h"
 #include "llvm/BinaryFormat/ELF.h"
+#include "llvm/IR/Constant.h"
 #include "llvm/IR/Constants.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/IR/Instructions.h"
 #include "llvm/IR/Module.h"
 #include "llvm/MC/MCContext.h"
 #include "llvm/MC/MCSectionELF.h"
 #include "llvm/MC/MCStreamer.h"
+#include "llvm/Support/ErrorHandling.h"
 
 using namespace llvm;
 using namespace std;
 
 namespace {
 
+class SerialiseInstructionException {
+private:
+  string S;
+
+public:
+  SerialiseInstructionException(string S) : S(S) {}
+  string &what() { return S; }
+};
+
 const char *SectionName = ".yk_ir";
 const uint32_t Magic = 0xedd5f00d;
 const uint32_t Version = 0;
 
 enum OpCode {
   Nop = 0,
+  Load,
+  Store,
+  Alloca,
+  Call,
+  GetElementPtr,
+  Branch,
+  ICmp,
+  BinaryOperator,
+  Ret,
+  UnimplementedInstruction = 255, // Will eventually be deleted.
+};
+
+enum OperandKind {
+  Constant = 0,
+  LocalVariable,
+  String,
+};
+
+enum TypeKind {
+  Integer = 0,
+  UnimplementedType = 255, // Will eventually be deleted.
 };
 
+string valueToString(Value *V) {
+  string S;
+  raw_string_ostream SS(S);
+  V->print(SS);
+  return S;
+}
+
+// Get the idex of an element in its parent container.
```

Comment:
This function is, if called frequently, potentially quite slow. I wondered if it's really necessary since, somewhere, I assume we once knew what the right index was (and then forgot it)? Not a big deal, but if it's easy to fix, it might be worth doing now, because threading through indices might be easier now than later.

Also sp/idex/index/ ?

Reply:
Yeah, it's not great, although it is a compile-time cost.

This also relates to the idea that our IR's instructions have a one-to-one relationship with LLVM's: an assumption that bothers me a bit since it means we can't exclude instructions during the lowering process (at least not without emitting NOPs in their place, which only bloats the IR). Any thoughts?

Would you be OK with me investigating that in a follow-up PR?

Reply:
(I will of course fix the typo now)

Reply:
> This also relates to the idea that our IR's instructions have a one-to-one relationship with LLVM's: an assumption that bothers me a bit since it means we can't exclude instructions during the lowering process (at least not without emitting NOPs in their place, which only bloats the IR). Any thoughts?

Why would missing instructions be a problem that causes us to put in place NOPs?

Reply:
Typo fixed in dd5424b160ceb1f8f820b9c238821377624bac19

Reply:
Ok, so a bit of background.

In LLVM's IR, an `Instruction` and a `Value` are one and the same. They use inheritance in a way such that a single pointer can describe both an operation, and the value that it generates (if any).

I find that elegant, and used a similar approach for our IR (although obviously we can't use pointers). In our IR, a local variable is described by the instruction that defines it. The way we identify the instruction is with a `(bb_idx, inst_idx)` pair.

I was lazy and queried the LLVM IR for these details when generating our IR. This means that there must be a 1-to-1 relationship between LLVM instruction indices and Yk IR instruction indices.

Suppose we have LLVM IR instructions in a block like:
```
bb0:
  call...
  add...
  sub...
```

And for whatever reason we want to exclude the `call` during lowering. By doing so we skew all the instruction indices by one, thus rendering the LLVM IR as an unworthy oracle for deciding instruction indices.

This can be fixed with a little state (hopefully in a separate PR). Like I said. I was being lazy, mostly because I'm struggling to bite off small chunks of functionality to PR.

Reply:
So you're saying the NOP thing is (if we use it!) a temporary hack that can go away? If so, fine with me: it's just that your original explanation kinda made it sound like it was fundamental.

Reply:
Yes and it would only be needed if we wanted to break the 1-to-1 mapping thing (which I think we should permit IMHO).

I have a list of PRs to do after this which fix/tidy design blunders and nits like this. I hope that's ok.